### PR TITLE
Make DPMultiheadAttention drop-in compatible with nn.MultiheadAttention

### LINKED ISSUE #123: DPMultiheadAttention is not drop-in compatible with nn.MultiheadAttention
## 🐛 Bug
DPMultiheadAttention is not drop-in compatible with nn.MultiheadAttention.

Opacus can only support certain "building block" layers (such as nn.Linear and nn.Conv2d), and layers written using those building blocks - in theory pretty much all of them, but in practice this leads to reimplementations. For example, while you can indeed write a LSTM using nn.Linear layers, the implementation we have in nn.LSTM doesn't do that (so that it can fuse ops and be faster). Whenever this happens, we reimplement the layer and make the update available in the Layers folder.

The aim of these reimplementations is that they all should be perfect drop-in replacements of their "normal PyTorch" counterparts, meaning that they should have exactly the same API, and exactly the same state_dict, so that users can train using the Opacus version, but then they can export the state_dict and load it for inference in a PyTorch environment that doesn't even have Opacus installed.

## Please reproduce using our [template Colab](https://colab.research.google.com/drive/1pCyN3_gN3WyhyQt8--eBPN5zVf2XZxiN?usp=sharing) and post here the link

Colab to repro this issue: https://colab.research.google.com/drive/1N4lqMmG21cROzx6qgJphTniMzcaX190C?usp=sharing

## To Reproduce

1. Instantiate a DPMultiheadAttention module
2. Instantiate a nn.MultiheadAttention with the same parameters
3. Their state_dicts should match!

## Additional context
There can be two reasons why this happens:

1. You have exactly the same matrices, but the names of the actual parameters are different due to how you organized your class.
2. You structured your layers differently so eg you have more/fewer matrices than the reference implementation

An example of 1 is the [DPLSTM](https://github.com/pytorch/opacus/blob/master/opacus/layers/dp_lstm.py). It's a big file, but essentially it has the same matrices as `nn.LSTM`, but they would be called something like `layers[0].ih` instead of `ih_0`. To fix this, we have an helper class named [ParamRenameModule](https://github.com/pytorch/opacus/blob/master/opacus/layers/param_rename.py) that allows you to specify a mapping. It's being used in DPLSTM, and it can help you here too.

In addition to this, here we also have Reason 2: we have a different number of matrices so we should align them. There are a few tricks that can help you here (this will teach you how `torch.nn.Module` works too!). Take a look at this code snippet to learn how PyTorch finds and registers parameters.

```python
class MyCoolModule(nn.Module):
    def __init__(self):
        super().__init__()  # starts the machinery to find submodules
        self.linear = nn.Linear(3, 2)  # this gets found and Torch will search for its params
        self.wrong_convs = [nn.Conv2d(i, 2*i) for i in range(5)]  # this DOES NOT get found!
        self.correct_convs = nn.ModuleList([nn.Conv2d(i, 2*i) for i in range(5)]) # This gets found!
        # Ditto with dicts and nn.ModuleDict

        # Note that all torch does is to recursively go inside and find nn.Parameters inside your 
        # submodules and calling self.register_parameter on those. So even if it doesn't find 
        # any params in self.wrong_convs, you CAN do this and it's exactly the same:
        for conv_module in self.wrong_convs:
            for pname, p in conv_module.named_parameters():
                registered_name = f"{conv_module}.{pname}"  # This is what nn.Module does, you can do whatever you want
                self.register_parameter(registered_name, p)
```

Remember: **Tensors are NOT Parameters!** Random tensors in a nn.Module won't get trained ever, they need to be boxed inside a nn.Parameter. Reference modules like `nn.Linear` do this for you already, so you don't generally notice this but this is how the sausage is made.

Here, you could use this behavior to your advantage by hiding things you don't want to be automatically registered and named by PyTorch, and instead manually registering the parameter with the name you want. This includes "tucking away" multiple matrices in a single one in a submodule, and hiding this submodule inside a list (list with single item) so that you can register it manually with a name that matches that of `nn.MultiheadAttention`.

### PR DESCRIPTION
**Pull request by Wei-1**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs change / refactoring / dependency upgrade

## Motivation and Context / Related issue

This PR is target to resolve #123 by having an additional re-naming mechanism to match the `state_dict` structure of `nn.MultiheadAttention`.

## How Has This Been Tested (if it applies)

Run `pytest` within the `opacus` folder:

```
============= 162 passed, 41 skipped, 4411 warnings in 243.18s (0:04:03) =============
```

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by Wei-1** [comment]
Remove for duplicated PR
